File: typescript/src/routes/auth.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import app from '../index';
import { authStore } from '../stores/auth';

describe('Auth Routes', () => {
  beforeEach(() => {
    // Clear all sessions before each test
    authStore.logout('');
    // @ts-ignore - accessing private property for testing
    authStore.sessions.clear();
  });

  describe('POST /auth/login', () => {
    it('should login with valid credentials', async () => {
      const res = await app.request('/auth/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          email: 'alice@example.com',
          password: 'password123',
        }),
      });

      expect(res.status).toBe(200);
      const body = await res.json();
      expect(body).toHaveProperty('accessToken');
      expect(body).toHaveProperty('tokenType', 'Bearer');
      expect(body).toHaveProperty('expiresIn', 86400);
      expect(body).toHaveProperty('user');
      expect(body.user).toMatchObject({
        id: '550e8400-e29b-41d4-a716-446655440001',
        email: 'alice@example.com',
        name: 'Alice Johnson',
      });
    });

    it('should fail with invalid credentials', async () => {
      const res = await app.request('/auth/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          email: 'alice@example.com',
          password: 'wrongpassword',
        }),
      });

      expect(res.status).toBe(401);
      const body = await res.json();
      expect(body.error).toMatchObject({
        code: 'UNAUTHORIZED',
        message: 'Invalid email or password',
      });
    });

    it('should fail with missing credentials', async () => {
      const res = await app.request('/auth/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          email: 'alice@example.com',
        }),
      });

      expect(res.status).toBe(400);
      const body = await res.json();
      expect(body.error).toMatchObject({
        code: 'BAD_REQUEST',
        message: 'Email and password are required',
      });
    });
  });

  describe('POST /auth/logout', () => {
    it('should logout with valid token', async () => {
      // First login to get a token
      const loginRes = await app.request('/auth/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          email: 'alice@example.com',
          password: 'password123',
        }),
      });
      const { accessToken } = await loginRes.json();

      // Then logout
      const res = await app.request('/auth/logout', {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${accessToken}`,
        },
      });

      expect(res.status).toBe(200);
      const body = await res.json();
      expect(body).toMatchObject({
        message: 'Logged out successfully',
      });

      // Verify token is invalidated
      const meRes = await app.request('/auth/me', {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${accessToken}`,
        },
      });
      expect(meRes.status).toBe(401);
    });

    it('should fail without authentication', async () => {
      const res = await app.request('/auth/logout', {
        method: 'POST',
      });

      expect(res.status).toBe(401);
      const body = await res.json();
      expect(body.error).toMatchObject({
        code: 'UNAUTHORIZED',
        message: 'Missing Authorization header',
      });
    });
  });

  describe('GET /auth/me', () => {
    it('should return current user with valid token', async () => {
      // First login to get a token
      const loginRes = await app.request('/auth/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          email: 'alice@example.com',
          password: 'password123',
        }),
      });
      const { accessToken } = await loginRes.json();

      // Get current user
      const res = await app.request('/auth/me', {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${accessToken}`,
        },
      });

      expect(res.status).toBe(200);
      const body = await res.json();
      expect(body).toHaveProperty('user');
      expect(body.user).toMatchObject({
        id: '550e8400-e29b-41d4-a716-446655440001',
        email: 'alice@example.com',
        name: 'Alice Johnson',
      });
    });

    it('should fail without authentication', async () => {
      const res = await app.request('/auth/me', {
        method: 'GET',
      });

      expect(res.status).toBe(401);
      const body = await res.json();
      expect(body.error).toMatchObject({
        code: 'UNAUTHORIZED',
        message: 'Missing Authorization header',
      });
    });

    it('should fail with invalid token', async () => {
      const res = await app.request('/auth/me', {
        method: 'GET',
        headers: {
          'Authorization': 'Bearer invalid-token',
        },
      });

      expect(res.status).toBe(401);
      const body = await res.json();
      expect(body.error).toMatchObject({
        code: 'UNAUTHORIZED',
        message: 'Invalid or expired token',
      });
    });
  });
});

File: go/internal/handlers/auth_test.go
package handlers

import (
	"bytes"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/blck-snwmn/hello-typespec/go/generated"
	authctx "github.com/blck-snwmn/hello-typespec/go/internal/auth"
	"github.com/blck-snwmn/hello-typespec/go/internal/storage"
	"github.com/blck-snwmn/hello-typespec/go/internal/store"
)

func TestAuthHandlers_Login(t *testing.T) {
	// Setup
	memoryStore := store.NewMemoryStore()
	authStore := storage.NewAuthStore()
	server := NewServer(memoryStore, authStore)

	// Test successful login with pre-configured user
	t.Run("successful login", func(t *testing.T) {
		loginReq := generated.LoginRequest{
			Email:    "alice@example.com",
			Password: "password123",
		}
		body, _ := json.Marshal(loginReq)

		req := httptest.NewRequest("POST", "/auth/login", bytes.NewReader(body))
		req.Header.Set("Content-Type", "application/json")
		w := httptest.NewRecorder()

		server.AuthServiceLogin(w, req)

		if w.Code != http.StatusOK {
			t.Errorf("expected status %d, got %d", http.StatusOK, w.Code)
		}

		var response generated.LoginResponse
		if err := json.NewDecoder(w.Body).Decode(&response); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		if response.AccessToken == "" {
			t.Error("expected access token, got empty")
		}
		if response.TokenType != "Bearer" {
			t.Errorf("expected token type Bearer, got %s", response.TokenType)
		}
		if response.User.Email != "alice@example.com" {
			t.Errorf("expected email alice@example.com, got %s", response.User.Email)
		}
	})

	// Test invalid email
	t.Run("invalid email", func(t *testing.T) {
		loginReq := generated.LoginRequest{
			Email:    "nonexistent@example.com",
			Password: "password",
		}
		body, _ := json.Marshal(loginReq)

		req := httptest.NewRequest("POST", "/auth/login", bytes.NewReader(body))
		req.Header.Set("Content-Type", "application/json")
		w := httptest.NewRecorder()

		server.AuthServiceLogin(w, req)

		if w.Code != http.StatusUnauthorized {
			t.Errorf("expected status %d, got %d", http.StatusUnauthorized, w.Code)
		}
	})

	// Test wrong password
	t.Run("wrong password", func(t *testing.T) {
		loginReq := generated.LoginRequest{
			Email:    "alice@example.com",
			Password: "wrongpassword",
		}
		body, _ := json.Marshal(loginReq)

		req := httptest.NewRequest("POST", "/auth/login", bytes.NewReader(body))
		req.Header.Set("Content-Type", "application/json")
		w := httptest.NewRecorder()

		server.AuthServiceLogin(w, req)

		if w.Code != http.StatusUnauthorized {
			t.Errorf("expected status %d, got %d", http.StatusUnauthorized, w.Code)
		}
	})
}

func TestAuthHandlers_Logout(t *testing.T) {
	// Setup
	memoryStore := store.NewMemoryStore()
	authStore := storage.NewAuthStore()
	server := NewServer(memoryStore, authStore)

	// Login to create a token
	session, err := authStore.Login("alice@example.com", "password123")
	if err != nil {
		t.Fatalf("Failed to create test session: %v", err)
	}
	token := session.Token

	// Test successful logout
	t.Run("successful logout", func(t *testing.T) {
		req := httptest.NewRequest("POST", "/auth/logout", nil)
		req.Header.Set("Authorization", "Bearer "+token)
		w := httptest.NewRecorder()

    server.AuthServiceLogout(w, req)

		if w.Code != http.StatusOK {
			t.Errorf("expected status %d, got %d", http.StatusOK, w.Code)
		}

		// Verify token is deleted by trying to validate it
		if _, err := authStore.ValidateToken(token); err == nil {
			t.Error("expected token to be deleted, but validation succeeded")
		}
	})

	// Test logout without token
	t.Run("logout without token", func(t *testing.T) {
		req := httptest.NewRequest("POST", "/auth/logout", nil)
		w := httptest.NewRecorder()

    server.AuthServiceLogout(w, req)

		if w.Code != http.StatusUnauthorized {
			t.Errorf("expected status %d, got %d", http.StatusUnauthorized, w.Code)
		}
	})
}

func TestAuthHandlers_GetCurrentUser(t *testing.T) {
	// Setup
	memoryStore := store.NewMemoryStore()
	authStore := storage.NewAuthStore()
	server := NewServer(memoryStore, authStore)

	// Login to create a token
	session, err := authStore.Login("alice@example.com", "password123")
	if err != nil {
		t.Fatalf("Failed to create test session: %v", err)
	}
	token := session.Token

	// Test successful get current user
	t.Run("successful get current user", func(t *testing.T) {
		req := httptest.NewRequest("GET", "/auth/me", nil)
		req.Header.Set("Authorization", "Bearer "+token)

		// Since GetCurrentUser expects user in context, we need to add it
		user := &storage.AuthUser{
			ID:    session.User.ID,
			Email: session.User.Email,
			Name:  session.User.Name,
		}
		ctx := authctx.WithUser(req.Context(), user)
		req = req.WithContext(ctx)

		w := httptest.NewRecorder()

    server.AuthServiceGetCurrentUser(w, req)

		if w.Code != http.StatusOK {
			t.Errorf("expected status %d, got %d", http.StatusOK, w.Code)
		}

		var authUser generated.AuthUser
		if err := json.NewDecoder(w.Body).Decode(&authUser); err != nil {
			t.Fatalf("failed to decode response: %v", err)
		}

		if authUser.Email != "alice@example.com" {
			t.Errorf("expected email alice@example.com, got %s", authUser.Email)
		}
		if authUser.Name != "Alice Johnson" {
			t.Errorf("expected name Alice Johnson, got %s", authUser.Name)
		}
	})

	// Test get current user without token (without context)
	t.Run("get current user without token", func(t *testing.T) {
		req := httptest.NewRequest("GET", "/auth/me", nil)
		w := httptest.NewRecorder()

		// This will panic because no user in context, so we need to handle it differently
		// In real usage, the auth middleware would return 401 before reaching this handler
		defer func() {
			if r := recover(); r != nil {
				// Expected panic when user is not in context
				w.WriteHeader(http.StatusUnauthorized)
			}
		}()

    server.AuthServiceGetCurrentUser(w, req)

		if w.Code != http.StatusUnauthorized && w.Code != 0 {
			t.Errorf("expected status %d or panic, got %d", http.StatusUnauthorized, w.Code)
		}
	})
}
